### PR TITLE
WIP: Add client builders and factory

### PR DESCRIPTION
**Pull request by wey1and**
Closes https://github.com/tarantool/cartridge-java/issues/110

### CODE REVIEW

File: src/main/java/io/tarantool/driver/clientbuilder/TarantoolProxyClientBuilder.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import io.tarantool.driver.ProxyTarantoolTupleClient;
+import io.tarantool.driver.api.TarantoolClient;
+import io.tarantool.driver.api.TarantoolResult;
+import io.tarantool.driver.api.tuple.TarantoolTuple;
+import io.tarantool.driver.proxy.ProxyOperationsMappingConfig;
+
+public interface TarantoolProxyClientBuilder
+        extends TarantoolDecoratedClientBuilder
+        <ProxyTarantoolTupleClient, TarantoolProxyClientBuilder,
+                TarantoolClient<TarantoolTuple, TarantoolResult<TarantoolTuple>>> {
```
Comment (ArtDu) [inline src/main/java/io/tarantool/driver/clientbuilder/TarantoolProxyClientBuilder.java:12]: Maybe that's more readable? (for all big generics)
```suggestion
<ProxyTarantoolTupleClient,
 TarantoolProxyClientBuilder,
 TarantoolClient<TarantoolTuple, TarantoolResult<TarantoolTuple>>> {
```

File: src/main/java/io/tarantool/driver/clientbuilder/TarantoolProxyClientBuilderImpl.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    public TarantoolProxyClientBuilderImpl() {
+        initBuilder();
+    }
+
+    @Override
+    protected void initBuilder() {
+        super.instance = this;
+    }
+}
```
Comment (ArtDu) [inline src/main/java/io/tarantool/driver/clientbuilder/TarantoolProxyClientBuilderImpl.java:44]: In all implementations, we have the same code, maybe there is an opportunity to take him higher to deduplicate

**Review by ArtDu — COMMENTED** [review]
Thanks for the patchset!
I just reviewed only the Builders commit.
See my comments inline and I have general comments regarding the whole patchset. I know that is the only WIP, but please add a description or explanation of changes to a commit message. This will make it easier to review.

File: src/main/java/io/tarantool/driver/clientbuilder/TarantoolRetryingClientBuilderImpl.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            return new RetryingTarantoolTupleClient(super.getDecoratedClient(), this.retryPolicyFactory);
+        }
+        return new RetryingTarantoolTupleClient(
+                super.getDecoratedClient(), this.retryPolicyFactory, this.executor);
+    }
+
+    @Override
+    protected void initBuilder() {
+        super.instance = this;
+    }
```
[the remark this answers is not in the document]
Comment (akudiyar) [reply src/main/java/io/tarantool/driver/clientbuilder/TarantoolRetryingClientBuilderImpl.java:43]: I also do not get this pattern. And seems that it is very bad: we modify some global state (a singleton instance in a class) with some concrete implementation with its own internal state. It breaks several design principles at once.

Consider using normal class instances created at their point of usage (in the Tarantool Client Factory).

File: src/main/java/io/tarantool/driver/clientfactory/AbstractTarantoolClientBuilderDecorator.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import io.tarantool.driver.TarantoolServerAddress;
+import io.tarantool.driver.auth.TarantoolCredentials;
+import io.tarantool.driver.clientbuilder.TarantoolClientBuilder;
+import io.tarantool.driver.clientbuilder.TarantoolClusterClientBuilder;
+import io.tarantool.driver.proxy.ProxyOperationsMappingConfig;
+import io.tarantool.driver.retry.RequestRetryPolicyFactory;
+
+import java.util.concurrent.Executor;
+
+public abstract class AbstractTarantoolClientBuilderDecorator<T, B extends TarantoolClientBuilder<T, B>>
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/clientfactory/AbstractTarantoolClientBuilderDecorator.java:15]: What's the meaning of this class, what problem does it solve?

File: src/main/java/io/tarantool/driver/clientfactory/TarantoolClientBuilders.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package io.tarantool.driver.clientfactory;
+
+public interface TarantoolClientBuilders {
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/clientfactory/TarantoolClientBuilders.java:3]: Some leftover?

File: src/main/java/io/tarantool/driver/clientfactory/TarantoolClientFactory.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package io.tarantool.driver.clientfactory;
+
+public interface TarantoolClientFactory {
+
+    TarantoolClientFactory INSTANCE = new TarantoolClientFactoryImpl();
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/clientfactory/TarantoolClientFactory.java:5]: Singleton factory instances are needed if they keep some state. In the other cases, static factory methods are enough.

File: src/main/java/io/tarantool/driver/clientfactory/TarantoolRetyingClientBuilderDecoratorImpl.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
+package io.tarantool.driver.clientfactory;
```
Comment (akudiyar) [inline src/main/java/io/tarantool/driver/clientfactory/TarantoolRetyingClientBuilderDecoratorImpl.java:1]: Why the factory and builder classes are not in the `api` package or in the root one, where the `TarantoolClient` resides?

The current driver structure implies the separation of classes into packages by domain, not by patterns they implement.

**Review by akudiyar — CHANGES_REQUESTED** [review]

**Pull request closed without merging**